File: learn-algo/leetcode/911/c++/911.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>
#include <math.h>

using namespace std;

class TopVotedCandidate {
public:
    vector<int> winnerIdByTime;
    vector<int> timeArr;

    TopVotedCandidate(vector<int> &persons, vector<int> &times) {
        int n = persons.size();
        vector<int> scores;
        timeArr = times;

        int scoreMax = 0;
        int winnerId = 0;

        scores.resize(n);

        for (int i = 0; i < n; ++i) {
            int id = persons[i];
            scores[id]++;

            if (scores[id] >= scoreMax) {
                scoreMax = scores[id];
                winnerId = id;

            }
            winnerIdByTime.push_back(winnerId);
        }
    }

    int q(int t) {
        int id = upper_bound(timeArr.begin(), timeArr.end(), t) - timeArr.begin() - 1;
        return winnerIdByTime[id];
    }
};

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate* obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj->q(t);
 */

int main() {
    vector<int> persons = {0, 0, 0, 0, 1};
    vector<int> times = {0, 6, 39, 52, 75};
    vector<int> queries = {45, 49, 59, 68, 42, 37, 99, 26, 78, 43};

    TopVotedCandidate *obj = new TopVotedCandidate(persons, times);

    for_each(obj->timeArr.begin(), obj->timeArr.end(), [](auto t) { cout << t << " "; });
    cout << endl;


    for (auto i = 0; i < obj->winnerIdByTime.size(); i++) {
        cout << obj->winnerIdByTime[i] << " ";
    }
    cout << endl;

    for (auto q: queries) {
        auto res = obj->q(q);
        cout << res << endl;
    }

    return 0;
}

File: learn/learn-algo/leetcode/911/go/911.go
package main

import (
	"fmt"
	"sort"
)

type TopVotedCandidate struct {
	winnerIdByTime []int
	timeArr        []int
}

func Constructor(persons []int, times []int) TopVotedCandidate {
	var (
		winnerIdByTime []int
		timeArr        []int
	)
	n := len(persons)
	var scores = make([]int, n)
	timeArr = times

	var scoreMax = 0
	var winnerId = 0

	for _, id := range persons {
		scores[id]++

		if scores[id] >= scoreMax {
			scoreMax = scores[id]
			winnerId = id
		}

		winnerIdByTime = append(winnerIdByTime, winnerId)
	}

	return TopVotedCandidate{winnerIdByTime, timeArr}
}

func (e *TopVotedCandidate) Q(t int) int {
	index := sort.Search(len(e.timeArr), func(i int) bool { return e.timeArr[i] > t })
	return e.winnerIdByTime[index-1]
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * obj := Constructor(persons, times);
 * param_1 := obj.Q(t);
 */

func main() {
	persons := []int{0, 0, 0, 0, 1}
	times := []int{0, 6, 39, 52, 75}
	//queries := []int{0, 5, 10, 15, 20, 25, 30}
	queries := []int{45, 49, 59, 68, 42, 37, 99, 26, 78, 43}

	obj := Constructor(persons, times)

	fmt.Println(obj.timeArr)
	fmt.Println(obj.winnerIdByTime)

	for _, q := range queries {
		res := obj.Q(q)
		fmt.Print(res, " ")
	}
}
